frontend/catalyst/api_extensions/quantum_passes.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains public API functions that provide control for the 
+user to input what MLIR compiler passes to run. 
+
+Currently, each pass has its own user-facing decorator. In the future, 
+a unified user interface for all the passes is necessary. 
+
+Note that the decorators do not need to modify the qnode in 
+any way. Its only purpose is to mark down the passes the user wants to 
+run on each qnode, and then generate the corresponding 
+transform.apply_apply_registered_pass in the lowered mlir.
+"""
+
+import pennylane as qml
+
+from catalyst.jax_primitives import apply_registered_pass_p, transform_named_sequence_p
+
+
+def inject_transform_named_sequence():
+    """
+    Inject a transform_named_sequence jax primitive.
+
+    This must be called when preprocessing the traced function in QJIT.capture(),
+    since to invoke -transform-interpreter, a transform_named_sequence primitive
+    must be in the jaxpr.
+    """
+
+    transform_named_sequence_p.bind()
+
+
+## API ##
+def cancel_inverses(fn=None):
+    """
+    The top-level ``catalyst.cancel_inverses`` decorator.
+
+    This decorator is always applied to a qnode, and it cancels two neighbouring self-inverse gates in the compiled mlir.
+
+
+    .. note::
+
+        Currently, only Hadamard gates are canceled.
+
+    .. note::
+
+        The qnode itself is not changed. In other words, circuit inspection tools such as ``qml.draw`` will still display the
+        neighbouring self-inverse gates. However, catalyst never executes the pennylane code directly. Instead catalyst
+        executes the compiled mlir, and these neighbouring self inverse gates are canceled in the compiled mlir.
+
+        To inspect the compiled mlir from Catalyst, use ``qjit(keep_intermediate=True)`` in the top-level ``qjit`` decorator.
+        This will create the intermediate mlir in a directory under the same directory where your python file is.
+        The cancel inverse happens at the stage ``QuantumCompilationPass``.
+
+    Args:
+        fn (Callable): a qml.QNode to run the cancel inverses transformation on
+
+    Returns:
+        The same qml.QNode.
```

Comment:
Technically the function of the qnode has been wrapped to inject the `apply_registered_pass` jax primitive, so this might not be accurate wording.

Reply:
Well it is the same `QNode` instance, so that is accurate. It is only the function that the QNode was wrapping that was modified